dagger/storageproofs/bls.nim

```diff
@@ -0,0 +1,390 @@
+
+  # generate the coefficient vector for combining sectors of a block: U
+  var ubase: seq[blst_scalar]
+  for i  in 0 ..< s :
+    let (u, ub) = rndP1()
+    t.u.add(u)
+    ubase.add(ub)
+   
+  #TODO: a better bytearray conversion of TauZero for the signature might be needed
+  let signature = sign(ssk.signkey, $t)
+  let tau = Tau(t: t, signature: signature.exportRaw())
+
+  #generate sigmas
+  var sigmas: seq[blst_p1]
+  for i in 0 ..< n :
+    sigmas.add(generateAuthenticator(i, s, t, ubase, file, ssk))
+
+  file.close()
+  result = (tau, sigmas)
+
+type QElement = object
+  I: int64
+  V: blst_scalar
+
+proc generateQuery*(tau: Tau, spk: PublicKey, l: int): seq[QElement] =
+  ## Generata a random BLS query of given sizxe
+  let n = tau.t.n # number of blocks
+
+  for i in 0 ..< l :
+    var q: QElement
+    q.I = Rng.instance.rand(n-1) #TODO: dedup
+    q.V = rndScalar() #TODO: fix range
+    result.add(q)
+
+proc generateProof*(q: openArray[QElement], authenticators: openArray[blst_p1], spk: PublicKey, s: int64, filename: string): (seq[blst_scalar], blst_p1) =
+  ## Generata BLS proofs for a given query
+  let file = open(filename)
+
+  var mu: seq[blst_scalar]
+  for j in 0 ..< s :
+    var muj: blst_fr
+    for qelem in q :
+      var x, v, sector: blst_fr
+      let sect = fromBytesBE(getSector(file, qelem.I, j, s))
+      sector.blst_fr_from_scalar(sect)
+      v.blst_fr_from_scalar(qelem.V)
+      x.blst_fr_mul(v, sector)
+      muj.blst_fr_add(muj, x)
+    var mujs: blst_scalar
+    mujs.blst_scalar_from_fr(muj)
+    mu.add(mujs)
+
+  var sigma: blst_p1
+  for qelem in q:
+    var prod: blst_p1
+    prod.blst_p1_mult(authenticators[qelem.I], qelem.V, 255)
+    sigma.blst_p1_add_or_double(sigma, prod)
+
+  file.close()
+  return (mu, sigma)
+
+proc pairing(a: blst_p1, b: blst_p2): blst_fp12 =
+  ## Calculate pairing G_1,G_2 -> G_T
+  var aa: blst_p1_affine
+  var bb: blst_p2_affine
+  blst_p1_to_affine(aa, a)
+  blst_p2_to_affine(bb, b)
+  var l: blst_fp12
+  blst_miller_loop(l, bb, aa)
+  blst_final_exp(result, l)
+
+proc verifyPairingsNaive(a1: blst_p1, a2: blst_p2, b1: blst_p1, b2: blst_p2) : bool =
+  let e1 = pairing(a1, a2)
+  let e2 = pairing(b1, b2)
+  return e1 == e2
+
+proc verifyPairingsNeg(a1: blst_p1, a2: blst_p2, b1: blst_p1, b2: blst_p2) : bool =
+  ## Faster pairing verification using 2 miller loops but ony one final exponentiation
+  ## based on https://github.com/benjaminion/c-kzg/blob/main/src/bls12_381.c
+  var
+    loop0, loop1, gt_point: blst_fp12
+    aa1, bb1: blst_p1_affine
+    aa2, bb2: blst_p2_affine
+
+  var a1neg = a1
+  blst_p1_cneg(a1neg, 1)
+
+  blst_p1_to_affine(aa1, a1neg)
+  blst_p1_to_affine(bb1, b1)
+  blst_p2_to_affine(aa2, a2)
+  blst_p2_to_affine(bb2, b2)
+
+  blst_miller_loop(loop0, aa2, aa1)
+  blst_miller_loop(loop1, bb2, bb1)
+
+  blst_fp12_mul(gt_point, loop0, loop1)
+  blst_final_exp(gt_point, gt_point)
+
+  return blst_fp12_is_one(gt_point).bool
```

Comment:
Alternatively the high-level primitives to do this is `blst_core_verify_pk_in_g1` (if pubkey in G1 and signature in G2)

https://github.com/status-im/nim-blscurve/blob/58729c1189871815e8d0c7b96957f0bac3cff9f7/blscurve/blst/blst_min_pubkey_sig_core.nim#L241-L283

Reply:
I'm using the lower level primitives just before, so it seems reasonable to me to keep it that way also for the final part.